File: desktop/lib/shannon/settings-v2.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
"use client";

import { getAccessToken, getAPIKey } from "@/lib/auth";

// =============================================================================
// Base API URL Helper
// =============================================================================

function getApiBaseUrl(): string {
    const isTauri = typeof window !== 'undefined' && '__TAURI__' in window;

    if (isTauri) {
        return (typeof window !== 'undefined' && window.__SHANNON_API_URL) || "";
    }

    return process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";
}

// =============================================================================
// Auth Headers Helper
// =============================================================================

function getAuthHeaders(): Record<string, string> {
    const headers: Record<string, string> = {};

    const apiKey = getAPIKey();
    if (apiKey) {
        headers["X-API-Key"] = apiKey;
        return headers;
    }

    const token = getAccessToken();
    if (token) {
        headers["Authorization"] = `Bearer ${token}`;
        return headers;
    }

    const userId = process.env.NEXT_PUBLIC_USER_ID;
    if (userId) {
        headers["X-User-Id"] = userId;
    }

    return headers;
}

// =============================================================================
// Enhanced Settings Types (Phase 8)
// =============================================================================

export interface AppSettings {
    id: string;
    providers: ProviderSettings[];
    models: ModelPreferences;
    appearance: AppearanceSettings;
    context: ContextSettings;
    knowledge: KnowledgeSettings;
    mcp: MCPSettings;
    advanced: AdvancedSettings;
    created_at: string;
    updated_at: string;
}

export interface ProviderSettings {
    provider: string;
    api_key: string;
    api_base?: string;
    enabled: boolean;
    default_model: string;
    config: Record<string, any>;
}

export interface ModelPreferences {
    default_quick_model: string;
    default_task_model: string;
    default_embedding_model: string;
    model_overrides: Record<string, string>;
    temperature_overrides: Record<string, number>;
}

export interface AppearanceSettings {
    theme: 'light' | 'dark' | 'auto' | { custom: string };
    custom_theme?: CustomTheme;
    language: string;
    font_family: string;
    font_size: number;
    message_density: string;
    sidebar_position: string;
}

export interface CustomTheme {
    name: string;
    colors: Record<string, string>;
    fonts: Record<string, string>;
    spacing: Record<string, string>;
}

export interface ContextSettings {
    max_context_tokens: number;
    retention_strategy: string;
    auto_summarize: boolean;
}

export interface KnowledgeSettings {
    chunking_strategy: string;
    chunk_size: number;
    chunk_overlap: number;
    embedding_provider: string;
    similarity_threshold: number;
    max_results: number;
}

export interface MCPSettings {
    enabled: boolean;
    servers: MCPServerConfig[];
    auto_discover: boolean;
}

export interface MCPServerConfig {
    name: string;
    command: string;
    args: string[];
    env: Record<string, string>;
    enabled: boolean;
}

export interface AdvancedSettings {
    debug_mode: boolean;
    telemetry_enabled: boolean;
    auto_update: boolean;
    concurrent_requests: number;
    request_timeout: number;
    experimental_features: boolean;
}

export type SettingsSection =
    | 'providers'
    | 'models'
    | 'appearance'
    | 'context'
    | 'knowledge'
    | 'mcp'
    | 'advanced';

// =============================================================================
// Settings V2 API Functions
// =============================================================================

/**
 * Get comprehensive application settings.
 */
export async function getAppSettings(): Promise<AppSettings> {
    const response = await fetch(`${getApiBaseUrl()}/api/v2/settings`, {
        method: "GET",
        headers: getAuthHeaders(),
    });

    if (!response.ok) {
        throw new Error(`Failed to get app settings: ${response.statusText}`);
    }

    return response.json();
}

/**
 * Update comprehensive application settings.
 */
export async function updateAppSettings(settings: AppSettings): Promise<void> {
    const response = await fetch(`${getApiBaseUrl()}/api/v2/settings`, {
        method: "PUT",
        headers: {
            "Content-Type": "application/json",
            ...getAuthHeaders(),
        },
        body: JSON.stringify({ settings }),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to update settings: ${response.statusText}`);
    }
}

/**
 * Update a specific settings section.
 */
export async function updateSettingsSection(
    section: SettingsSection,
    value: any
): Promise<void> {
    const response = await fetch(`${getApiBaseUrl()}/api/v2/settings/${section}`, {
        method: "PUT",
        headers: {
            "Content-Type": "application/json",
            ...getAuthHeaders(),
        },
        body: JSON.stringify(value),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to update ${section}: ${response.statusText}`);
    }
}

/**
 * Export settings as YAML.
 */
export async function exportSettings(): Promise<string> {
    const response = await fetch(`${getApiBaseUrl()}/api/v2/settings/export`, {
        method: "GET",
        headers: getAuthHeaders(),
    });

    if (!response.ok) {
        throw new Error(`Failed to export settings: ${response.statusText}`);
    }

    return response.text();
}

/**
 * Import settings from YAML.
 */
export async function importSettings(yaml: string): Promise<void> {
    const response = await fetch(`${getApiBaseUrl()}/api/v2/settings/import`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            ...getAuthHeaders(),
        },
        body: JSON.stringify({ yaml }),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to import settings: ${response.statusText}`);
    }
}

// =============================================================================
// Default Settings
// =============================================================================

export const DEFAULT_APP_SETTINGS: AppSettings = {
    id: "default",
    providers: [],
    models: {
        default_quick_model: "gpt-4o-mini",
        default_task_model: "gpt-4o",
        default_embedding_model: "text-embedding-3-small",
        model_overrides: {},
        temperature_overrides: {},
    },
    appearance: {
        theme: "auto",
        language: "en",
        font_family: "Inter",
        font_size: 14,
        message_density: "normal",
        sidebar_position: "left",
    },
    context: {
        max_context_tokens: 128000,
        retention_strategy: "sliding_window",
        auto_summarize: true,
    },
    knowledge: {
        chunking_strategy: "recursive",
        chunk_size: 1000,
        chunk_overlap: 200,
        embedding_provider: "openai",
        similarity_threshold: 0.7,
        max_results: 5,
    },
    mcp: {
        enabled: true,
        servers: [],
        auto_discover: true,
    },
    advanced: {
        debug_mode: false,
        telemetry_enabled: false,
        auto_update: true,
        concurrent_requests: 5,
        request_timeout: 300,
        experimental_features: false,
    },
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
};

// =============================================================================
// Legacy Settings API Functions (for backward compatibility)
// =============================================================================

export interface UserSetting {
    user_id: string;
    setting_key: string;
    setting_value: string;
    setting_type: string;
    encrypted: boolean;
    created_at: string;
    updated_at: string;
}

export interface SetSettingRequest {
    key: string;
    value: string;
    setting_type?: string;
    encrypted?: boolean;
}

export interface ApiKeyInfo {
    provider: string;
    is_configured: boolean;
    masked_key: string | null;
    is_active: boolean;
    last_used_at: string | null;
    created_at: string | null;
}

export interface SetApiKeyResponse {
    provider: string;
    masked_key: string;
    message: string;
}

export async function getAllSettings(): Promise<UserSetting[]> {
    const response = await fetch(`${getApiBaseUrl()}/api/v1/settings`, {
        method: "GET",
        headers: getAuthHeaders(),
    });

    if (!response.ok) {
        throw new Error(`Failed to get settings: ${response.statusText}`);
    }

    return response.json();
}

export async function getSetting(key: string): Promise<UserSetting> {
    const response = await fetch(`${getApiBaseUrl()}/api/v1/settings/${encodeURIComponent(key)}`, {
        method: "GET",
        headers: getAuthHeaders(),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to get setting: ${response.statusText}`);
    }

    return response.json();
}

export async function setSetting(
    key: string,
    value: string,
    type: string = "string",
    encrypted: boolean = false
): Promise<UserSetting> {
    const response = await fetch(`${getApiBaseUrl()}/api/v1/settings`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            ...getAuthHeaders(),
        },
        body: JSON.stringify({
            key,
            value,
            setting_type: type,
            encrypted,
        }),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to set setting: ${response.statusText}`);
    }

    await response.json();

    try {
        return await getSetting(key);
    } catch {
        return {
            user_id: "",
            setting_key: key,
            setting_value: value,
            setting_type: type,
            encrypted,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString(),
        };
    }
}

export async function deleteSetting(key: string): Promise<void> {
    const response = await fetch(`${getApiBaseUrl()}/api/v1/settings/${encodeURIComponent(key)}`, {
        method: "DELETE",
        headers: getAuthHeaders(),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to delete setting: ${response.statusText}`);
    }
}

export async function listApiKeys(): Promise<ApiKeyInfo[]> {
    const response = await fetch(`${getApiBaseUrl()}/api/v1/settings/api-keys`, {
        method: "GET",
        headers: getAuthHeaders(),
    });

    if (!response.ok) {
        throw new Error(`Failed to list API keys: ${response.statusText}`);
    }

    return response.json();
}

export async function setApiKey(provider: string, apiKey: string): Promise<SetApiKeyResponse> {
    const response = await fetch(`${getApiBaseUrl()}/api/v1/settings/api-keys/${encodeURIComponent(provider)}`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            ...getAuthHeaders(),
        },
        body: JSON.stringify({
            api_key: apiKey,
        }),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to set API key: ${response.statusText}`);
    }

    return response.json();
}

export async function deleteApiKey(provider: string): Promise<void> {
    const response = await fetch(`${getApiBaseUrl()}/api/v1/settings/api-keys/${encodeURIComponent(provider)}`, {
        method: "DELETE",
        headers: getAuthHeaders(),
    });

    if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `Failed to delete API key: ${response.statusText}`);
    }
}

File: rust/shannon-api/src/database/settings_v2.rs
//! Enhanced settings system with comprehensive configuration support.
//!
//! This module extends the basic settings with structured application settings
//! including providers, models, appearance, context, knowledge, MCP, and advanced settings.

use crate::database::encryption::KeyManager;
use crate::database::hybrid::HybridBackend;
use anyhow::{Context, Result};
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Comprehensive application settings.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppSettings {
    /// User ID or "default" for system-wide settings
    pub id: String,
    /// Provider configurations
    pub providers: Vec<ProviderSettings>,
    /// Model preferences
    pub models: ModelPreferences,
    /// Appearance and theme settings
    pub appearance: AppearanceSettings,
    /// Context management settings (from Phase 5)
    pub context: ContextSettings,
    /// Knowledge base settings (from Phase 7)
    pub knowledge: KnowledgeSettings,
    /// MCP server settings (from Phase 6)
    pub mcp: MCPSettings,
    /// Advanced settings
    pub advanced: AdvancedSettings,
    /// When the settings were created
    pub created_at: DateTime<Utc>,
    /// When the settings were last updated
    pub updated_at: DateTime<Utc>,
}

/// Provider configuration for LLM services.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderSettings {
    /// Provider identifier (openai, anthropic, google, groq, xai)
    pub provider: String,
    /// API key (encrypted in storage)
    pub api_key: String,
    /// Custom API base URL (optional)
    pub api_base: Option<String>,
    /// Whether provider is enabled
    pub enabled: bool,
    /// Default model for this provider
    pub default_model: String,
    /// Additional provider-specific configuration
    pub config: HashMap<String, serde_json::Value>,
}

/// Model selection preferences.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelPreferences {
    /// Default model for quick queries
    pub default_quick_model: String,
    /// Default model for complex tasks
    pub default_task_model: String,
    /// Default embedding model
    pub default_embedding_model: String,
    /// Per-capability model overrides
    pub model_overrides: HashMap<String, String>,
    /// Model-specific temperature settings
    pub temperature_overrides: HashMap<String, f32>,
}

impl Default for ModelPreferences {
    fn default() -> Self {
        Self {
            default_quick_model: "gpt-4o-mini".to_string(),
            default_task_model: "gpt-4o".to_string(),
            default_embedding_model: "text-embedding-3-small".to_string(),
            model_overrides: HashMap::new(),
            temperature_overrides: HashMap::new(),
        }
    }
}

/// Appearance and theme configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppearanceSettings {
    /// Theme mode
    pub theme: ThemeMode,
    /// Custom theme configuration
    pub custom_theme: Option<CustomTheme>,
    /// Language code (en, es, fr, etc.)
    pub language: String,
    /// Font family
    pub font_family: String,
    /// Font size in pixels
    pub font_size: u8,
    /// Message density (compact, normal, comfortable)
    pub message_density: String,
    /// Sidebar position (left, right)
    pub sidebar_position: String,
}

impl Default for AppearanceSettings {
    fn default() -> Self {
        Self {
            theme: ThemeMode::Auto,
            custom_theme: None,
            language: "en".to_string(),
            font_family: "Inter".to_string(),
            font_size: 14,
            message_density: "normal".to_string(),
            sidebar_position: "left".to_string(),
        }
    }
}

/// Theme mode selection.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ThemeMode {
    /// Light theme
    Light,
    /// Dark theme
    Dark,
    /// System preference
    Auto,
    /// Custom theme by name
    Custom(String),
}

/// Custom theme definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CustomTheme {
    /// Theme name
    pub name: String,
    /// Color palette
    pub colors: HashMap<String, String>,
    /// Font definitions
    pub fonts: HashMap<String, String>,
    /// Spacing definitions
    pub spacing: HashMap<String, String>,
}

/// Context management settings (from Phase 5).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextSettings {
    /// Maximum context window size
    pub max_context_tokens: usize,
    /// Context retention strategy
    pub retention_strategy: String,
    /// Whether to enable auto-summarization
    pub auto_summarize: bool,
}

impl Default for ContextSettings {
    fn default() -> Self {
        Self {
            max_context_tokens: 128_000,
            retention_strategy: "sliding_window".to_string(),
            auto_summarize: true,
        }
    }
}

/// Knowledge base settings (from Phase 7).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeSettings {
    /// Default chunking strategy
    pub chunking_strategy: String,
    /// Chunk size for documents
    pub chunk_size: usize,
    /// Chunk overlap
    pub chunk_overlap: usize,
    /// Default embedding provider
    pub embedding_provider: String,
    /// Similarity threshold for retrieval
    pub similarity_threshold: f32,
    /// Maximum results to retrieve
    pub max_results: usize,
}

impl Default for KnowledgeSettings {
    fn default() -> Self {
        Self {
            chunking_strategy: "recursive".to_string(),
            chunk_size: 1000,
            chunk_overlap: 200,
            embedding_provider: "openai".to_string(),
            similarity_threshold: 0.7,
            max_results: 5,
        }
    }
}

/// MCP server settings (from Phase 6).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MCPSettings {
    /// Whether MCP is enabled
    pub enabled: bool,
    /// Configured MCP servers
    pub servers: Vec<MCPServerConfig>,
    /// Auto-discover MCP servers
    pub auto_discover: bool,
}

impl Default for MCPSettings {
    fn default() -> Self {
        Self {
            enabled: true,
            servers: Vec::new(),
            auto_discover: true,
        }
    }
}

/// MCP server configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MCPServerConfig {
    /// Server name
    pub name: String,
    /// Server command
    pub command: String,
    /// Server arguments
    pub args: Vec<String>,
    /// Environment variables
    pub env: HashMap<String, String>,
    /// Whether server is enabled
    pub enabled: bool,
}

/// Advanced system settings.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdvancedSettings {
    /// Enable debug mode
    pub debug_mode: bool,
    /// Enable telemetry
    pub telemetry_enabled: bool,
    /// Enable auto-update
    pub auto_update: bool,
    /// Maximum concurrent requests
    pub concurrent_requests: u8,
    /// Request timeout in seconds
    pub request_timeout: u64,
    /// Enable experimental features
    pub experimental_features: bool,
}

impl Default for AdvancedSettings {
    fn default() -> Self {
        Self {
            debug_mode: false,
            telemetry_enabled: false,
            auto_update: true,
            concurrent_requests: 5,
            request_timeout: 300,
            experimental_features: false,
        }
    }
}

impl Default for AppSettings {
    fn default() -> Self {
        Self {
            id: "default".to_string(),
            providers: Vec::new(),
            models: ModelPreferences::default(),
            appearance: AppearanceSettings::default(),
            context: ContextSettings::default(),
            knowledge: KnowledgeSettings::default(),
            mcp: MCPSettings::default(),
            advanced: AdvancedSettings::default(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }
}

/// Repository trait for enhanced settings management.
#[async_trait]
pub trait SettingsV2Repository: Send + Sync {
    /// Get comprehensive settings for a user.
    ///
    /// # Errors
    ///
    /// Returns an error if the database query fails.
    async fn get_app_settings(&self, user_id: &str) -> Result<AppSettings>;

    /// Save comprehensive settings for a user.
    ///
    /// # Errors
    ///
    /// Returns an error if the database write fails.
    async fn save_app_settings(&self, settings: &AppSettings) -> Result<()>;

    /// Export settings as YAML.
    ///
    /// # Errors
    ///
    /// Returns an error if serialization fails.
    async fn export_settings(&self, user_id: &str) -> Result<String>;

    /// Import settings from YAML.
    ///
    /// # Errors
    ///
    /// Returns an error if deserialization or save fails.
    async fn import_settings(&self, user_id: &str, yaml: &str) -> Result<()>;

    /// Update a specific settings section.
    ///
    /// # Errors
    ///
    /// Returns an error if the update fails.
    async fn update_section(
        &self,
        user_id: &str,
        section: SettingsSection,
        value: serde_json::Value,
    ) -> Result<()>;
}

/// Settings section identifier for partial updates.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SettingsSection {
    /// Provider settings
    Providers,
    /// Model preferences
    Models,
    /// Appearance settings
    Appearance,
    /// Context settings
    Context,
    /// Knowledge settings
    Knowledge,
    /// MCP settings
    Mcp,
    /// Advanced settings
    Advanced,
}

#[async_trait]
impl SettingsV2Repository for HybridBackend {
    async fn get_app_settings(&self, user_id: &str) -> Result<AppSettings> {
        let user_id = user_id.to_string();
        let sqlite = self.sqlite.clone();

        tokio::task::spawn_blocking(move || -> Result<AppSettings> {
            let guard = sqlite.lock().unwrap();
            let conn = guard
                .as_ref()
                .ok_or_else(|| anyhow::anyhow!("SQLite not initialized"))?;

            // Try to load from app_settings table
            let mut stmt = conn.prepare(
                "SELECT settings_json, created_at, updated_at
                 FROM app_settings WHERE user_id = ?1",
            )?;

            let mut rows = stmt.query(params![user_id])?;

            if let Some(row) = rows.next()? {
                let json_str: String = row.get(0)?;
                let mut settings: AppSettings =
                    serde_json::from_str(&json_str).context("Failed to parse settings JSON")?;

                settings.id = user_id;
                settings.created_at = parse_datetime(row.get::<_, String>(1)?);
                settings.updated_at = parse_datetime(row.get::<_, String>(2)?);

                Ok(settings)
            } else {
                // Return default settings if none exist
                let mut settings = AppSettings::default();
                settings.id = user_id;
                Ok(settings)
            }
        })
        .await
        .context("Tokio spawn_blocking failed")?
    }

    async fn save_app_settings(&self, settings: &AppSettings) -> Result<()> {
        let user_id = settings.id.clone();
        let settings_json =
            serde_json::to_string(settings).context("Failed to serialize settings")?;
        let sqlite = self.sqlite.clone();
        let now = Utc::now().to_rfc3339();

        tokio::task::spawn_blocking(move || -> Result<()> {
            let guard = sqlite.lock().unwrap();
            let conn = guard
                .as_ref()
                .ok_or_else(|| anyhow::anyhow!("SQLite not initialized"))?;

            conn.execute(
                "INSERT INTO app_settings (user_id, settings_json, created_at, updated_at)
                 VALUES (?1, ?2, ?3, ?3)
                 ON CONFLICT(user_id) DO UPDATE SET
                   settings_json = excluded.settings_json,
                   updated_at = excluded.updated_at",
                params![user_id, settings_json, now],
            )?;
            Ok(())
        })
        .await
        .context("Tokio spawn_blocking failed")?
    }

    async fn export_settings(&self, user_id: &str) -> Result<String> {
        let settings = self.get_app_settings(user_id).await?;

        // Mask sensitive data before export
        let mut export_settings = settings.clone();
        for provider in &mut export_settings.providers {
            provider.api_key = "***REDACTED***".to_string();
        }

        serde_yaml::to_string(&export_settings).context("Failed to serialize settings to YAML")
    }

    async fn import_settings(&self, user_id: &str, yaml: &str) -> Result<()> {
        let mut settings: AppSettings =
            serde_yaml::from_str(yaml).context("Failed to parse settings YAML")?;

        settings.id = user_id.to_string();
        settings.updated_at = Utc::now();

        // Don't import redacted API keys
        settings.providers.retain(|p| p.api_key != "***REDACTED***");

        self.save_app_settings(&settings).await
    }

    async fn update_section(
        &self,
        user_id: &str,
        section: SettingsSection,
        value: serde_json::Value,
    ) -> Result<()> {
        let mut settings = self.get_app_settings(user_id).await?;

        match section {
            SettingsSection::Providers => {
                settings.providers =
                    serde_json::from_value(value).context("Invalid providers data")?;
            }
            SettingsSection::Models => {
                settings.models = serde_json::from_value(value).context("Invalid models data")?;
            }
            SettingsSection::Appearance => {
                settings.appearance =
                    serde_json::from_value(value).context("Invalid appearance data")?;
            }
            SettingsSection::Context => {
                settings.context = serde_json::from_value(value).context("Invalid context data")?;
            }
            SettingsSection::Knowledge => {
                settings.knowledge =
                    serde_json::from_value(value).context("Invalid knowledge data")?;
            }
            SettingsSection::Mcp => {
                settings.mcp = serde_json::from_value(value).context("Invalid MCP data")?;
            }
            SettingsSection::Advanced => {
                settings.advanced =
                    serde_json::from_value(value).context("Invalid advanced data")?;
            }
        }

        settings.updated_at = Utc::now();
        self.save_app_settings(&settings).await
    }
}

/// Parse datetime from RFC3339 string.
fn parse_datetime(value: String) -> DateTime<Utc> {
    DateTime::parse_from_rfc3339(&value)
        .map(|dt| dt.with_timezone(&Utc))
        .unwrap_or_else(|_| Utc::now())
}

/// SQL schema for app_settings table.
pub const APP_SETTINGS_SCHEMA: &str = r#"
CREATE TABLE IF NOT EXISTS app_settings (
    user_id TEXT PRIMARY KEY,
    settings_json TEXT NOT NULL,
    created_at TEXT DEFAULT (datetime('now')),
    updated_at TEXT DEFAULT (datetime('now'))
);
"#;
